src/core/processing/qgsprocessingutils.cpp

```diff
@@ -172,3 +173,29 @@ QString QgsProcessingUtils::normalizeLayerSource( const QString &source )
   return normalized.trimmed();
 }
 
+QgsFeatureIterator QgsProcessingUtils::getFeatures( QgsVectorLayer *layer, const QgsProcessingContext &context, const QgsFeatureRequest &request )
+{
+  bool useSelection = context.flags() & QgsProcessingContext::UseSelection && layer->selectedFeatureCount() > 0;
+
+  QgsFeatureRequest req( request );
+  req.setInvalidGeometryCheck( context.invalidGeometryCheck() );
+  req.setInvalidGeometryCallback( context.invalidGeometryCallback() );
+  if ( useSelection )
+  {
+    return layer->selectedFeaturesIterator( req );
+  }
+  else
+  {
+    return layer->getFeatures( req );
+  }
+}
+
+long QgsProcessingUtils::featureCount( QgsVectorLayer *layer, const QgsProcessingContext &context )
+{
+  bool useSelection = context.flags() & QgsProcessingContext::UseSelection && layer->selectedFeatureCount() > 0;
```

Comment:
Could we leave the discussion about this "feature" out of this PR? Because:

- it's the current behaviour, and has been for a while, so changing it should first be discussed on the dev list
- I don't want this PR held up with that discussion - I'd rather get this merged and then we can refine/change based on the other discussion

For now this is a direct copy of the existing logic